File: MyLeetCodeAnswers/C++/SwapNodesInPairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *swapPairs(ListNode *head) {
        if (head == NULL || head->next == NULL) return head;
        ListNode *newHead = head->next;
        ListNode *pre = NULL;
        ListNode *cur = head;
        ListNode *ne = NULL;
        while (cur != NULL && cur->next != NULL) {
            ne = cur->next->next;
            if (pre != NULL) {
                pre->next = cur->next;
            }
            cur->next->next = cur;
            cur->next = ne;
            pre = cur;
            cur = ne;
        }
        return newHead;
    }
};

/*可以设置一个假头，辅助遍历！*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *swapPairs(ListNode *head) {
        if (head == NULL || head->next == NULL) return head;
        ListNode *dummy = new ListNode(-1);
        dummy->next = head;
        ListNode *pre = dummy;
        ListNode *cur = head;
        ListNode *ne = NULL;
        while (cur != NULL && cur->next != NULL) {
            ne = cur->next->next;
            pre->next = cur->next;
            cur->next->next = cur;
            cur->next = ne;
            pre = cur;
            cur = ne;
        }
        return dummy->next;
    }
};

File: MyLeetCodeAnswers/java/SwapNodesInPairs.java
/**
 * Source   : https://leetcode.com/problems/merge-two-sorted-lists/
 * Author   : Wang Meng
 * Date     : 2016-03-19
 * <p>
 * Given a linked list, swap every two adjacent nodes and return its head.
 * <p>
 * For example,
 * Given 1->2->3->4, you should return the list as 2->1->4->3.
 * <p>
 * Your algorithm should use only constant space. You may not modify the values in the list, only nodes itself can be changed.
 */
public class SwapNodesInPairs {
    // recursive solution
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode ret = head.next;
        ListNode p = head.next.next;
        head.next.next = head;
        head.next = swapPairs(p);
        return ret;
    }
}
